scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/MultiSourceKeyGroupReader.java

```diff
@@ -0,0 +1,234 @@
+          // "  "    "   "     "   "    "          rejected by a preceding source, don't emit.
+          // if this is the first source for this key group, emit if either the source settings or
+          // the min key say we should.
+          boolean emitKeyGroup =
+              (acceptKeyGroup == AcceptKeyGroup.ACCEPT)
+                  || ((acceptKeyGroup == AcceptKeyGroup.UNSET)
+                      && (src.emitByDefault || emitBasedOnMinKeyBucketing));
+
+          final Iterator<Object> keyGroupIterator = (Iterator<Object>) src.currentValue();
+          if (emitKeyGroup) {
+            acceptKeyGroup = AcceptKeyGroup.ACCEPT;
+            // data must be eagerly materialized if requested or if there is a predicate
+            boolean materialize = materializeKeyGroup || src.predicate.isPresent();
+            int outputIndex = resultSchema.getIndex(src.tupleTag);
+
+            if (!materialize) {
+              // lazy data iterator
+              valueMap.set(
+                  outputIndex,
+                  new SortedBucketSource.TraversableOnceIterable<>(
+                      Iterators.transform(
+                          keyGroupIterator,
+                          (value) -> {
+                            runningKeyGroupSize++;
+                            return value;
+                          })));
+
+            } else {
+              // eagerly materialize this iterator and apply the predicate to each value
+              // this must be eager because the predicate can operate on the entire collection
+              final List<Object> values = (List<Object>) valueMap.get(outputIndex);
+              final SortedBucketSource.Predicate<Object> predicate =
+                  src.predicate
+                      .map(value -> (SortedBucketSource.Predicate<Object>) value)
+                      .orElseGet(() -> (xs, x) -> true);
+              keyGroupIterator.forEachRemaining(
+                  v -> {
+                    if ((predicate).apply(values, v)) {
+                      values.add(v);
+                      runningKeyGroupSize++;
+                    }
+                  });
+            }
+          } else {
+            acceptKeyGroup = AcceptKeyGroup.REJECT;
+            // skip key but still have to exhaust iterator
+            keyGroupIterator.forEachRemaining(value -> {});
+          }
+        }
+      }
+
+      if (acceptKeyGroup == AcceptKeyGroup.ACCEPT) {
+        final KV<byte[], CoGbkResult> next =
+            KV.of(minKey, CoGbkResultUtil.newCoGbkResult(resultSchema, valueMap));
+        try {
+          // new head found, we're done
+          head =
+              Optional.of(
+                  KV.of(keyCoder.decode(new ByteArrayInputStream(next.getKey())), next.getValue()));
+          break;
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to decode key group", e);
+        }
+      }
+    }
+  }
+
+  private static class BucketedInputSource<K, V> {
```

Comment:
I'm a bit torn on whether it's clearer to have this as a static class or just inline the method code in `MultiSourceKeyGroupReader`. Since we already have `BucketedInput`, `SortedBucketSource`, & `KeyGroupIterator` I'm worried about added complexity/name overlap :) wdyt?

Reply:
I refactored this code out of its original location so the code in the `MultiSourceKeyGroupReader` would be more intelligible. It's private so hopefully not polluting the namespace. Definitely open to suggestions for a better name!